### DISCUSSION TITLE
How do I prevent company by sorting by substring?

### DISCUSSION THREAD

**Discussion by sethen** [company-mode/company-mode#1246] [Q&A] [answered]
Every time I type something company wants to sort by substring when I want a literal match.  As you can see here, I would expect "console" to be at the top.  How do I change the sorting for this?  I don't see it in the documentation anywhere.
![Screenshot from 2021-10-22 15-37-57](https://user-images.githubusercontent.com/1493530/138513591-d0e34383-4ced-4107-82ca-41f755ad3fc6.png)

**Comment by dgutov** [comment]
This matching strategy is decided by the used backend. Sorting -- either by it as well, or by the contents of `company-transformers`.

You can find out which backend is used by typing `M-x company-diag`.

**Reply by sethen** [reply]
Here's my code

```elisp
(require 'use-package)

(use-package company
  :bind (:map company-active-map
			  ("M-j" . 'company-select-next)
			  ("M-k" . 'company-select-previous)
			  ("M-l" . 'company-complete))
  :commands
  (company-complete company-select-next company-select-previous global-company-mode)
  :config
  (setq company-transformers '(company-sort-by-backend-importance)
		company-idle-delay 0
		company-minimum-prefix-length 1)
  :init
  (global-company-mode))

(provide 'sethen-company)
```

I am using orderless and it seems that when orderless is in the mix it start sorting it all weird.  I only have 2 backends enabled `company-capf` and `company-keywords`.  Is there some sort of integration setting that I am supposed to use with orderless?

**Comment by sethen** [comment, marked as answer]
I think I figured it out... Setting `completion-styles` in orderless to `'(basic orderless)` does exactly what I want.

**Reply by dgutov** [reply]
That's a good solution: the `basic` style is used first; it tries to find completions which are straight prefix matches. If so, those are returned, and `orderless` is not used.

The `basic` style doesn't do any sorting. `orderless` does.

**Comment by sethen** [comment]
I am actually coming back to this because the original solution I provided _does_ work but it will mess with company results when you group them.  Instead I have run orderless on the mini-buffer hook so orderless will leave company results alone.

**Reply by dgutov** [reply]
`completion-styles` only affect what `company-capf` does. The effect should be reliable enough, but might be weird when combined with grouping (exactly because those styles don't affect other backends).

There might be bugs in the grouping logic too, of course.

**Reply by sethen** [reply]
@dgutov Actually now I am getting results that are sorting by substrings again and not literal completions even with orderless turned off... So this isn't the answer.  It looks the same as the initial screenshot I posted... Why is company trying to sort by substrings as opposed to literal completion when `completion-styles` has `basic`?

**Reply by sethen** [reply]
@dgutov So setting company-backends like this

```elisp
(set (make-local-variable 'company-backends) '((company-yasnippet company-capf company-keywords)))
```
Is causing issues in the sorting.  Getting rid of this line makes company sort again the way I think it should but now I can't add `company-yasnippet`??

**Reply by yugaego** [reply]
The follow-up discussion is https://github.com/company-mode/company-mode/discussions/1249.
